include/deal.II/lac/la_parallel_vector.h

```diff
@@ -175,13 +175,58 @@ namespace LinearAlgebra
      * fail in some circumstances. Therefore, it is strongly recommended to
      * not rely on this class to automatically detect the unsupported case.
      *
-     * @author Katharina Kormann, Martin Kronbichler, 2010, 2011, 2016
+     * <h4>CUDA support</h4>
+     *
+     * This vector class supports two different memory spaces: Host and CUDA. By
+     * default, the memory space is Host and all the data are allocated on the
+     * CPU. When the memory space is CUDA all the data is allocated on the GPU.
+     * The operations on the vector are performed on the chosen memory space. *
+     * From the host, there are two methods to access the elements of the Vector
+     * when using the CUDA memory space: <ul>
+     * <li> use get_values()
+     * <code>
+     * Vector<double, CUDA> vector(local_range, comm);
+     * double* vector_dev = vector.get_values();
+     * std::vector<double> vector_host(local_range.n_elements(), 1.);
+     * Utilities::CUDA::copy_to_dev(vector_host, vector_dev);
+     * </code>
+     * <li> use import()
+     * <code>
+     * Vector<double, CUDA> vector(local_range, comm);
+     * ReadWriteVector<double> rw_vector(local_range);
+     * for (auto & val : rw_vector)
+     *   val = 1.;
+     * vector.import(rw_vector, VectorOperations::insert);
+     * </code>
+     * </ul>
+     * The import method is a lot safer and will perform an MPI communication if
+     * necessary. Since an MPI communication may be performed, import needs to
+     * be called on all the processors.
+     *
+     * @note Vector using CUDA memory space are not instantiated by default so
+     * to use them it is necessary to include the la_parallel_vector.templates.h
+     * header. For the same reason, the read_write_vector.templates.h header
+     * must be included to use import.
+     * @note By default, all the ranks will try to access the device 0. This is
+     * fine is if you have one rank per node \a and one gpu per node. If you
+     * have multiple GPUs on one node, we need each process to access a
+     * different GPU. If each node has the same number of GPUs, this can be done
+     * as follows:
+     * <code> int n_devices = 0; cudaGetDeviceCount(&n_devices); int
+     * device_id = my_rank % n_devices;
+     * cudaSetDevice(device_id);
+     * </code>
+     * @see CUDAWrappers
+     *
+     * @author Katharina Kormann, Martin Kronbichler, Bruno Turcksin 2010, 2011,
+     * 2016, 2018
      */
-    template <typename Number>
+    template <typename Number, typename MemorySpace = Host>
```

Comment:
Instead of using a tag type here, you could also create an enum, say
```
  enum class MemorySpace {
    host, cuda
  };
```
and then use the following as a template:
```
  template <typename Number, MemorySpace memory_space>
  class Vector;
```
This may be simpler because it avoids people instantiating the class with template arguments that don't work.

Reply:
I have put some `static_assert` so it won't compile if the memory space is not `Host` or `CUDA` but I can use an enum if you prefer.

Reply:
I don't really have a preference. The `enum` solution clearly enumerates all possible choices for the argument, without the need to document this anywhere or the need to create a bunch of otherwise empty classes. 

The class approach you have makes sense if you eventually plan to put member functions into these classes. This would then be the "policy" design pattern, in the style of the `allocator` arguments to the C++ STL classes that all have an allocator that describes how to get and release memory.